### PR TITLE
Add Custom Evervault Errors

### PR DESCRIPTION
**Pull request by lfarrel6**
Added custom evervault errors for the sdk
- `BadRequest` -> 400s in `evervault.set`
- `Unauthorized` -> 401s in `evervault.get` and `evervault.set`
- `DataNotFound` -> 404s in `evervault.get`
- `EncryptionError` -> errors in `evervault.encrypt`
- `DecryptionError` -> errors in `evervault.decrypt`
- `EvervaultError` -> unexpected errors from the sdk

The custom errors are exposed through `evervault.errors`

Added tests to ensure these errors are thrown as expected

### CODE REVIEW

File: errors.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/** @format */
+
+class DomainError extends Error {
```
Comment (jonnyom) [inline errors.js:3]: Can we call this `EvervaultError`? And everything extends from that!

File: index.js [hunk 2]
```diff
@@ -112,40 +119,52 @@ module.exports = {
...
     })
-      .then((res) => res.json())
-      .then((result) => {
-        if (!result) {
-          throw new Error('Unable to get data');
+      .then((res) => {
+        if (res.ok) {
+          return res.json();
+        }
+        if (res.status === 400) {
```
Comment (jonnyom) [inline index.js:152]: Would this be neater in a switch?

File: errors.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  constructor(message, origin, data) {
+    super(`[${origin}] ${message} ${JSON.stringify(data)}`);
+    this.data = data;
+  }
+}
+
+class EvervaultDataNotFound extends DomainError {
+  constructor(origin, data) {
+    super('The requested data could not be found.', origin, data);
```
Comment (jonnyom) [inline errors.js:12]: This is a weird nit.

I'm not sure if we should have full stops in here or not. For some reason it reads weirdly to me, but that could just be me too.

File: errors.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/** @format */
```
Comment (eoinboylan) [inline errors.js:1]: Its generally bad practice to put the company name in the Class name it has no meaning and its people now its our sdk.

Comment (eoinboylan) [reply errors.js:1]: Leaving this cause i approved by accident

File: errors.js (hunk 3 above)
Comment (lfarrel6) [reply errors.js:12]: Yeah that's fair enough, I dont think most standard errors have it so I'll remove the '.'

File: errors.js (hunk 4 above)
Comment (jonnyom) [reply errors.js:1]: Hmm taking a look at Stripe's API they put some in.
https://stripe.com/docs/api/errors/handling
I'd personally lean towards having the base class be EvervaultError (not SdkError - there are lots of SDKs, which one failed?) And have the others inherit from that.
In two minds about keeping evervault in the others. Just thinking about what would surface in something like Sentry. Helping people debug is a big deal imo

Comment (lfarrel6) [reply errors.js:1]: Yeah I agree, I think the SdkError feels weird. Prepending them with Evervault would make debugging easier for people IMO

They could have some function that calls evervault and another library side by side (say making a http request with the returned evervault data from a get) and throwing an `EvervaultError` is much clearer in those situations than say a `BadRequestError`

Comment (jonnyom) [reply errors.js:1]: Yeah exactly. I think letting people know what's causing the error is important

Comment (eoinboylan) [reply errors.js:1]: Thats stripes API not SDK, you know what throwing the error its tells you in the stack trace?

Comment (eoinboylan) [reply errors.js:1]: I'm not going to die on this hill, i've been taught not to litter your class names with a company name.

Comment (jonnyom) [reply errors.js:1]: They're errors from the SDK though. I'm not aware of that convention personally 🤷

Comment (lfarrel6) [reply errors.js:1]: So I've updated the errors to account for this discussion..

The root error is called an `EvervaultError` incase it is thrown unexpectedly, but every other error doesn't contain Evervault.

The thrown errors themselves are prepended with an origin which references where in the sdk the error occurred and will provide the level of visibility that we should give!

An error message might say: `[evervault.get] The requested data could not be found {requested: <dataToGet>}`

**Pull request merged**